doc/man/txt2bxl.1
.\" pcb-rnd - manual
.\" Copyright (C) 2016 Tibor 'Igor2' Palinkas
.\" 
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\" 
.\" Contact: pcb-rnd-man[removethis]@igor2.repo.hu
.TH txt2bxl 1 2020-04-18 "" "pcb-rnd manual"
.SH NAME
txt2bxl - encode a BXL file provided in plain text and save binary
.SH SYNPOSIS
.nf
.sp
\fBtxt2bxl \fIoutfile.bxl\fB &lt; \fIinfile.txt\fB
.fi
.SH DECSRIPTION

.BR txt2bxl
reads and encodes a text bxl file in its standard input and saves the resulting binary bxl file on the file name specified in the command line argument.


Changelog
pcb-rnd 2.2.1 (r30749)
~~~~~~~~~~~~~~~~~~~~~~
	[autoplace]
		-Fix: replace the totally broken code for "smd vs. thru-hole" query
		-Fix: do not allow overlap of subc bbox on the primary layer (fixes the problem that was caused by not allowing padstack/term overlap only)

	[brave]
		-Change: invert clipbatch - Gabriel probably would have reported already if anything broke, consider bacth clipping safe

	[build]
		-Fix: do not compile librnd for make clean
		-Fix: use target CC for genlist aux object compilation
		-Del: build system special casing for core action registration with auto-generated lists - since core actions are only a small part of all actions by now, there is no point in special casing; remove PCB_REGISTER_ACTIONS_FUNC() and PCB_REGISTER_ACTIONS_CALL() (split API for semi-automated action registration) - there is only one, manual API now
		-Split: hooks for librnd to make common infra reusable for sch-rnd
		-Move: calculate generic cflags before src/Makefile, so they can be included in Makefile.conf

	[cam]
		-Cleanup: make the error message of 'failed to produce any content' more clear about the reason (no layer group)
		-Add: separate global option for okempty-content and okempty-group

	[conf]
		-Add: new drc const subtree
		-Add: make sure centrally defined new drc consts are kept in sync with the legacy version unless they are overdefined

	[core]
		-Fix: undo: batch poly clipping to speed things up
		-Fix: wireframe_draw, thin_draw, thin_draw_poly and check_planes is communicated to the draw code through xform instead of global conf - export doesn't have to automatically reproduce GUI behavior on this
		-Fix: do not draw padstack shapes in check_planes mode - padstacks are not considered "planes", not even if they are of polygon shape
		-Fix: hide_names, show_solder_side, invis_other_groups, black_current_group  draw is communicated to the draw code through xform instead of global conf - export doesn't have to automatically reproduce GUI behavior on this
		-Fix: loadfrom(netlist, ...) prints an error message if none of the import plugins could handle the file
		-Fix: do not try to remove unlisted subcircuits on import that have the NONETLIST flag set
		-Fix: data_mirror batches all low level undo into a single serial
		-Fix: cut-to-buffer object removal is undoable
		-Fix: rat anchor resolver on polygon shall use forced poly clipping to make sure the endpoint geometry calculation can be done
		-Fix: don't crash on search if it ends up on a polygon that got cleared out of existnece
		-Fix: message action: don't do invalid memory reads when argument can not be converted to string
		-Fix: ChangeName() shouldn't crash if there's no parent subc on the object being changed
		-Fix: in compat mode flag conversion accepts octagon flag on padstacks
		-Fix: potentially uninitialized variable in ChangeName()
		-Fix: segfault on {e t} on no-object
		-Fix: Select() redraws the objects that changed flags (or the whole screen)
		-Fix: arc angles edit: final/output delta sanitization so it's never larger than 360 deg
		-Fix: arc angles undo shall reclip sorrunding polygon
		-Fix: configure default mode (tool id) to -1 for invalid; any other tool id is unsafe as tool IDs are dynamically allocated on plugin load in random order
		-Fix: tool: arrow tool selects copy/move/line tool by name instead of hardwired tool IDs
		-Fix: tool: arrow selection: increase undo serial after new selection is made so it is not undone together with the operation that follows
		-Fix: rats_patch API bug: global var declaration shall be extern (potential compilation error)
		-Fix: rat line drawing shouldn't snap to -6;-6 mil as a special case, it should use the normal grid snapping as per default
		-Del: crosshair range - do not try to restrict objects into the drawing area
		-Cleanup: move standard tools out to a (core) plugin called tool_std
		-Cleanup: reference tools by name, not ID, where possible, in preparation of fixed tool ID removal
		-Cleanup: rat layer tool enforcement is event based so the tool code does not need to know anything about the application specific tool logics
		-Cleanup: move tool logic from crosshair to tool_logic.c
		-Cleanup: generalize action execution from file, using librnd's new call instead of local implementation
		-Cleanup: remove static tool IDs from standard tools
		-Move: tool_note is really crosshair_note - it is application specific, move it from generic tool infra to crosshair
		-Move: hid crosshair notify function to librnd - it is a generic hid API + event wrapper
		-Change: API CHANGE: plug_import support_prio can not have a FILE * argument because the meaning of args[] is implementation-specific - do not open the first arg as file centrally either
		-Add: object implementation: gfx
		-Add: tool API: escape() callback so each tool can handle escape locally
		-Add: API CHANGE: extend the import plug API for multiple file names
		-Add: API CHANGE: split plug import api's new single_file and the idea of "all arguments are filenames" so the API is generic enough for external commands (e.g. make)
		-Add: plug_import API: mark plugins that execute user specified external commands (import_net_cmd)
		-Add: plug_import: utility call to look up importer by name
		-Add: new conf nodes for the footprint import placement, dispersion and footprint removal
		-Add: import excess part removal option: view list
		-Add: conf setting for silently creating boards that are named on the command line (useful in batch/scripting to create initial boards)
		-Add: Undo() commands: FreezeSerial, UnFreezeSerial and IncSerial so scripts can batch undo operations
		-Add: make line, arc, text, poly, padstack, subcircuit mirroring undoable
		-Add: xform: flag_color bit to indicate whether selected/found/warn color should be drawn; pstk, arc, line, poly, text rendering ignores flag colors when intructed

util/bxl2txt/txt2bxl.c
/*
 *                            COPYRIGHT
 *
 *  pcb-rnd, interactive printed circuit board design
 *
 *  BXL IO plugin - file format write, encoder
 *  pcb-rnd Copyright (C) 2020 Tibor 'Igor2' Palinkas
 *  (Supported by NLnet NGI0 PET Fund in 2020)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 *  Contact:
 *    Project page: http://repo.hu/projects/pcb-rnd
 *    lead developer: http://repo.hu/projects/pcb-rnd/contact.html
 *    mailing list: pcb-rnd (at) list.repo.hu (send "subscribe")
 */

/* test program: read plain text on stdin, encode to bxl to the named file */

#include <stdio.h>
#include "bxl_decode.h"

void out(FILE *f, hdecode_t *ctx, int len)
{
	int n;
	for(n = 0; n < len; n++)
		fputc(ctx->out[n], f);
}

int main(int argc, char *argv[])
{
	hdecode_t ctx;
	int inch, len, fd;
	FILE *f;

	if (argc < 2) {
		fprintf(stderr, "Need an output file name\n");
		return 1;
	}

	f = fopen(argv[1], "wb");
	if (f == NULL) {
		fprintf(stderr, "Failed to open %s for write\n", argv[1]);
		return 1;
	}

	pcb_bxl_encode_init(&ctx);

	while((inch = fgetc(stdin)) != EOF)
		out(f, &ctx, pcb_bxl_encode_char(&ctx, inch));
	out(f, &ctx, pcb_bxl_encode_eof(&ctx));

	fseek(f, 0, SEEK_SET);
	out(f, &ctx, pcb_bxl_encode_len(&ctx));

	return 0;
}
